x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java

```diff
@@ -24,29 +28,39 @@
  */
 class ProgressListenableActionFuture extends PlainActionFuture<Long> {
 
+    private static final Logger logger = LogManager.getLogger(ProgressListenableActionFuture.class);
+
     private record PositionAndListener(long position, ActionListener<Long> listener) {}
 
-    protected final long start;
-    protected final long end;
+    final long start;
+    final long end;
 
-    // modified under 'this' mutex
-    private volatile List<PositionAndListener> listeners;
-    protected volatile long progress;
+    /**
+     * A consumer that accepts progress made by this {@link ProgressListenableActionFuture}. The consumer is called before listeners are
+     * notified of the updated progress value in {@link #onProgress(long)}. The consumer can be called with out-of-order progress values.
+     */
```

Comment:
Can we call out that the consumer is not called when `progress == end`? I find it important and a bit counter-intuitive. So it's helpful to explicitly note it down.

Reply:
Done in 16052d4fa5cf003d0243518f41c08c51a8002314